#47 category 모듈 적용

Sung-jin opened the pull request:
related #47 
* 오디언스 모듈화 적용 및 다른 공통 모듈 적용, 버그 수정 진행하였습니다.

Sung-jin committed:
* category 모듈 적용

- 같은 키의 header 로 라우팅 체크시, 먼저 체크되는 설정만 먹는 현상 수정
- auth 모듈의 path 들이 auth 라우팅 설정에서 동작하지 않는 현상 수정

---
TODO - 현재 라우팅 설정에 따른 uri 프록시 설정이 동작하지 않고있으며, 수정 필요

Sung-jin committed:
* 모듈별 라우팅 설정 proxy 적용 안되던 현상 수정

gwanhyeon commented on growing-shop-api/audience/src/main/java/com/example/growingshopaudience/api/CategoryController.java:
```diff
@@ -28,14 +28,14 @@ public ResponseEntity<CategoryResponse.CategoryRes> create(@RequestBody Category
 
     @DeleteMapping("/{id}")
-    public ResponseEntity delete(@PathVariable Long id) {
+    public ResponseEntity<Object> delete(@PathVariable Long id) {
```
삭제시 ResponseEntity<Void> 형태로 HttpsStatus의 return 상태값을 no_content 204로 응답하는것이 조금더 명확하다고 생각하는데 어떻게 생각하시나요? 

자원 삭제 요청을 했고 이 요청이 유효하니 서버는 해당 자원을 삭제했다. 더 이상 응답할 컨텐츠가 없기 때문에 컨텐츠가 없는 204로 응답한다 라는 글을 보게되어서 한번 코멘트 남겨보았습니다.

제네릭 타입으로 처리 안하시고 Object타입으로 타입을 명시한부분은 👍

Sung-jin replied:
삭제 후 응답에 별다른 값이 없는 경우가 많은데, 해당 부분을 204 로 처리하는게 좋겠네요.
수정하겠습니다. 👍

gwanhyeon commented on growing-shop-api/audience/src/main/java/com/example/growingshopaudience/config/security/CategoryAuthorizeMatchers.java:
```diff
@@ -0,0 +1,22 @@
+    ) {
+        return registry
+                .requestMatchers("/h2-console/**").permitAll()
+                .requestMatchers(HttpMethod.GET, "/categories").permitAll()
```
HttpMethod.GET 방식만 허용된 것으로 보여지는데, 삭제 Post, Delete도 추가되어야 하는게 아닌지 의도가 궁금합니다.

Sung-jin replied:
기본적으로 security 모듈을 사용하는 모든 모듈의 api 는 화이트리스트 형태로 동작합니다.
즉, 모든 엔드포인트에 대해서는 인증을 통해서만 접근이 가능하며, 위와 같이 별도로 등록된 엔드포인트의 설정에 대해서만 인증 없이도 접근 가능한 형태입니다.

카테고리의 경우 모든 카테고리를 ui 에서 유저와 상관없이 사용될 수 있다고 생각하여 public 으로 열어두었으나, 비즈니스 요구사항에 따라 변경될 수 있습니다.

---

추가적으로, auth 모듈의 config/routing 의 라우팅 설정에도 위와 같은 형태의 허용되는 엔드포인트에 대한 설정이 있습니다.
모든 api 요청에 대해 auth 모듈을 통해서만 접근이 가능하도록 설정할 예정이며, auth 모듈에서 요청의 설정에 따라 라우팅을 하도록 할 예정입니다.
이때, routing 하위의 설정에 따라 로그인 여부를 체크하고, 해당 유저를 리다이렉트된 서비스에서 사용할 수 있도록 redis 에 설정하여 공통으로 사용되는 등의 형태로 활용됩니다.

이는 auth 모듈과 각 모듈의 엔드포인트 설정의 중복이 발생하고 있으며, 이는 추후 고도화를 통해 하나의 모듈 또는 기능에서 담당하도록 수정할 수 있게 고민할 예정입니다.

gwanhyeon reviewed (approved):
카테고리 모듈 궁금한 부분 코멘트 남겨놓았습니다.

코틀린과 자바를 동시에 모듈화, 프로파일 분리하여 적용하신부분 👍 

나머지 application.yml에 사용되고 있는 값들도 보안측면에서 secret.yml 하나 만들어서 value값 적용시켜도 좋을 것 같습니다!

Sung-jin committed:
* 주문 모듈 분리 및 라우팅 적용 (#49)

* auth 모듈의 auth api 는 spring security 로 인증되도록 수정

Sung-jin committed:
* 리뷰 적용

- delete api 의 응답 status 를 OK -> NO_CONTENT 로 변경
- ResponseEntity 의 원사사용을 Object 타입으로 변경
